### PR TITLE
[Python] Hugging Face pipeline support

### PR DESCRIPTION
**Pull request by riteshghorse**
This PR adds Hugging Face pipeline support to RunInference by adding a new model handler for it. Build on top of #26632 

[Example Job with Question Answering Pipeline on Dataflow](https://pantheon.corp.google.com/dataflow/jobs/us-central1/2023-08-01_12_16_27-9044957391141813799;graphView=0?pageState=(%22dfTime%22:(%22s%22:%222023-08-01T19:16:28.294Z%22,%22e%22:%222023-08-01T19:19:44.586Z%22))&project=google.com:clouddfe)

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by riteshghorse** [comment]
Run Python 3.8 PostCommit

**Comment by riteshghorse** [comment]
Run Python 3.11 PostCommit

File: sdks/python/apache_beam/ml/inference/huggingface_inference.py [hunk 1]
```diff
@@ -483,6 +499,129 @@ def share_model_across_processes(self) -> bool:
...
+    **Note:** To specify which device to use (CPU/GPU),
+    use the load_model_args with key-value as you would do in the usual
+    Hugging Face pipeline. Ex: load_model_args={'device':0})
+
+    Example Usage model::
+      pcoll | RunInference(HuggingFacePipelineModelHandler(
+        task="fill-mask"))
+
+    Args:
+      task (str): task supported by HuggingFace Pipelines.
```
Comment (liferoad) [inline sdks/python/apache_beam/ml/inference/huggingface_inference.py:550]: str or enum?

Comment (riteshghorse) [reply sdks/python/apache_beam/ml/inference/huggingface_inference.py:550]: It is a str. There are a lot of tasks to be defined under enum for us https://huggingface.co/docs/transformers/main/main_classes/pipelines

Comment (liferoad) [reply sdks/python/apache_beam/ml/inference/huggingface_inference.py:550]: Better provide the link for users to check what tasks are: https://huggingface.co/transformers/v4.12.5/_modules/transformers/pipelines.html#pipeline has the current list. We could support both enum and str.

Comment (riteshghorse) [reply sdks/python/apache_beam/ml/inference/huggingface_inference.py:550]: sounds good

**Comment by riteshghorse** [comment]
Run Python 3.8 PostCommit

**Comment by riteshghorse** [comment]
hugging face integration tests passed on 3.8 PostCommit 
image

**Comment by riteshghorse** [comment]
Run Python 3.8 PostCommit

**Comment by riteshghorse** [comment]
PTAL, ready for review. Hugging Face integration tests passed. Some other tests unrelated to this PR are failing

**Comment by riteshghorse** [comment]
assign next reviewer

**Comment by riteshghorse** [comment]
assign to next reviewer

**Comment by riteshghorse** [comment]
R: @damccorm

File: sdks/python/apache_beam/ml/inference/huggingface_inference.py [hunk 2]
```diff
@@ -483,6 +533,130 @@ def share_model_across_processes(self) -> bool:
...
+class HuggingFacePipelineModelHandler(ModelHandler[str,
+                                                   PredictionResult,
+                                                   Pipeline]):
+  def __init__(
+      self,
+      task: Union[str, PipelineTask] = "",
+      model=None,
+      *,
+      inference_fn: PipelineInferenceFn = _default_pipeline_inference_fn,
+      load_model_args: Optional[Dict[str, Any]] = None,
```
Comment (damccorm) [inline sdks/python/apache_beam/ml/inference/huggingface_inference.py:566]: Should this be `load_pipeline_args` per https://docs.google.com/document/d/107T71QHMxvu67PyG-mkHwRhm7W2-07jihnqUMz-yuEM/edit

File: sdks/python/apache_beam/ml/inference/huggingface_inference.py [hunk 3]
```diff
@@ -483,6 +533,130 @@ def share_model_across_processes(self) -> bool:
...
+    predictions: Union[Iterable, Dict[Any, Iterable]],
+    model_id: Optional[str] = None,
+) -> Iterable[PredictionResult]:
+  return [
+      PredictionResult(x, y, model_id) for x, y in zip(batch, [predictions])
+  ]
+
+
+def _default_pipeline_inference_fn(
+    batch, model, inference_args) -> Iterable[PredictionResult]:
```
Comment (damccorm) [inline sdks/python/apache_beam/ml/inference/huggingface_inference.py:552]: Nit: Could we change model to pipeline in this function and others? I think that will help for readability, and some might try to use this as an example of what to do.

File: sdks/python/apache_beam/ml/inference/huggingface_inference.py [hunk 4]
```diff
@@ -483,6 +533,130 @@ def share_model_across_processes(self) -> bool:
...
+        inference function.
+    Returns:
+      An Iterable of type PredictionResult.
+    """
+    inference_args = {} if not inference_args else inference_args
+    predictions = self._inference_fn(batch, model, inference_args)
+    return _convert_to_result(batch, predictions)
+
+  def update_model_path(self, model_path: Optional[str] = None):
+    self._model = model_path if model_path else self._model
```
Comment (damccorm) [inline sdks/python/apache_beam/ml/inference/huggingface_inference.py:650]: I'm trying to figure out what model updates mean in the context of hugging face pipelines. I imagine not too many people would do it, but I think leaving it makes sense as is. If they do updates, its unlikely to be a task update. Could you add a brief docstring to this function explaining what model updates mean.

File: sdks/python/apache_beam/ml/inference/huggingface_inference_it_test.py [hunk 5]
```diff
@@ -74,6 +75,37 @@ def test_hf_language_modeling(self):
       predicted_predicted_text = predictions_dict[text]
       self.assertEqual(actual_predicted_text, predicted_predicted_text)
 
+  def test_hf_pipeline(self):
+    test_pipeline = TestPipeline(is_integration_test=True)
+    # Path to text file containing some questions and context
+    input_file = 'gs://apache-beam-ml/datasets/custom/questions.txt'
+    output_file_dir = 'gs://apache-beam-ml/hf/testing/predictions'
```
Comment (damccorm) [inline sdks/python/apache_beam/ml/inference/huggingface_inference_it_test.py:82]: Will this break if the underlying model changes? Could we pin to a single revision?

Comment (damccorm) [reply sdks/python/apache_beam/ml/inference/huggingface_inference_it_test.py:82]: Ooh, actually same question about the existing tests

File: sdks/python/apache_beam/examples/inference/huggingface_question_answering.py [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+
+def preprocess(text):
+  if len(text.strip()) > 0:
+    question, context = text.split(';')
+    yield (question, context)
+
+
+def create_squad_example(text):
+  question, context = text
+  yield question, QuestionAnsweringPipeline.create_sample(question, context)
```
Comment (damccorm) [inline sdks/python/apache_beam/examples/inference/huggingface_question_answering.py:66]: Could you add a docstring to this function explaining what it is doing? Probably good to do that for the other preprocessing function too.

**Review by damccorm — COMMENTED** [review]
Left some comments, but the core pieces look good to me

File: sdks/python/apache_beam/ml/inference/huggingface_inference_it_test.py (hunk 5 above)
Comment (riteshghorse) [reply sdks/python/apache_beam/ml/inference/huggingface_inference_it_test.py:82]: Sounds good. yeah internal changes would corrupt our model. Pinning it to `4.30.0` for `huggingface_tests_requirements.txt`.

Comment (damccorm) [reply sdks/python/apache_beam/ml/inference/huggingface_inference_it_test.py:82]: I actually mean the model itself that the pipeline is built on top of, e.g. if they update from v2 to v3 - https://huggingface.co/deepset/roberta-base-squad2/commit/deedc3e42208524e0df3d9149d1f26aa6934f05f

I guess that's _that_ likely to happen, but we could just pin to v2 or even a commit sha with the revision parameter - https://huggingface.co/docs/transformers/main_classes/pipelines#transformers.pipeline.revision

Comment (riteshghorse) [reply sdks/python/apache_beam/ml/inference/huggingface_inference_it_test.py:82]: Got it, updated the test, thanks!

**Comment by damccorm** [comment]
Run Python 3.8 PostCommit

**Review by damccorm — APPROVED** [review]
LGTM once all suites pass

**Comment by riteshghorse** [comment]
Tests passed but not reflected here. Opened https://github.com/apache/beam/issues/27808 for the failing PreCommits. It is not related to this change.

Merging this PR!

**Comment by liferoad** [comment]
Can you resolve my comments?

**Comment by riteshghorse** [comment]
I've added the enum you suggested. Are there any comments you didn't publish yet accidentally?

**Comment by liferoad** [comment]
Interesting. I probably forgot to submit them. Sorry about that.

File: sdks/python/apache_beam/ml/inference/huggingface_inference_it_test.py [hunk 7]
```diff
@@ -74,6 +75,37 @@ def test_hf_language_modeling(self):
...
+
+    predictions_dict = {}
+    for prediction in predictions:
+      text, predicted_text = prediction.split(';')
+      predictions_dict[text] = predicted_text.strip()
+
+    for actual in actuals:
+      text, actual_predicted_text = actual.split(';')
+      predicted_predicted_text = predictions_dict[text]
+      self.assertEqual(actual_predicted_text, predicted_predicted_text)
```
[the remark this answers is not in the document]
Comment (damccorm) [reply sdks/python/apache_beam/ml/inference/huggingface_inference_it_test.py:108]: I had this question, and we addressed by setting a model revision to use

[the remark this answers is not in the document]
Comment (liferoad) [reply sdks/python/apache_beam/ml/inference/huggingface_inference_it_test.py:108]: Make sense. Thanks!

File: sdks/python/apache_beam/ml/inference/huggingface_inference.py [hunk 8]
```diff
@@ -483,6 +499,129 @@ def share_model_across_processes(self) -> bool:
...
+  ]
+
+
+def _default_pipeline_inference_fn(
+    batch, model, inference_args) -> Iterable[PredictionResult]:
+  predicitons = model(batch, **inference_args)
+  return predicitons
+
+
+class HuggingFacePipelineModelHandler(ModelHandler[str,
```
[the remark this answers is not in the document]
Comment (riteshghorse) [reply sdks/python/apache_beam/ml/inference/huggingface_inference.py:562]: For Keyed, we can just wrap this inside `KeyedModelHandler`. Added the update_model_path in later commits.

**Comment by riteshghorse** [comment]
Thought so! I wanted to get this in for 2.50 so Danny reviewed it as you were OOO.

**Comment by liferoad** [comment]
Thanks, feel free to resolve my comments.

**Pull request merged**